### DISCUSSION TITLE
Replace `null` with `none`, and add a `some` type.

### DISCUSSION THREAD

**Discussion by zkat** [kdl-org/kdl#37] [Ideas]
Because of the whole [`billion dollar mistake`](https://en.wikipedia.org/wiki/Null_pointer#History) thing, I was thinking we could allow special syntax for KDL to specifically mark any Values as "nullable"/"Option" types. These can well be treated on de/serialization as actual nullable types in languages like JavaScript that support it, or they may be translated to `Option` types in those that handle "absence" that way.

My main reasoning for introducing this as explicit syntax is to allow the distinction to be made _explicit_ by those who want to, such as certain target languages (like Rust!) or schema languages intended to be used with KDL, which may want to do static verification of a KDL schema.

My proposed syntax is this:

```kdl
my-node <> // None
my-node <1> // Some(1)
my-node <<1>> // Some(Some(1))
my-node <<>> // Some(None)
```

My intention is that in languages with null, which want to keep that logic, will represent `<>` as `null`, and `<1>` as a completely unwrapped `1`. Same thing for any level of wrapping, so `<<<<1>>>>` would become `1` in those languages. Likewise, `<<<>>>` would just become a simple `null`. Languages or libraries which choose to go the "safe" route, though, will be able to represent these as `Some(Some(Some(1)))` and such.

What do y'all think?

**Comment by lulu-berlin** [comment]
Angle brackets have the advantage of being shorter than `some(value)`, but still the question remains of whether you want the serialization format to enforce the difference between a required value and an optional value.

So, if `my-node` expects an optional number and the data holds a number, would it still be allowed to serialize it as
```
my-node 123
```
or only
```
my-node <123>
```
?

What about omitting the value altogether? Is that something you want to support, or rather require it to be explicit?

Is
```
my-node
```
equal to
```
my-node <>
```
?

**Reply by lulu-berlin** [reply]
What about other cases where a value can have different types? Would you introduce some `enum`-like functionality, or just let it be duck-typed?
```
my-node StringOrNumber::Number(123)
my-node StringOrNumber::String("abc")
```

or
```
my-node 123
my-node "abc"
```

I think that these design questions are related, and it boils down to: explicit enforcing of the model into the data, vs. loose data that is structured when deserialized.

**Reply by zkat** [reply]
In your first example: I note that `KDL` on its own is "dynamically" typed, and whether a known-"none-able" value is serialized as `123` or `<123>` is up to the particular implementation. However, this distinction can be enforced by adding a schema language on top of KDL (something I assume will happen if KDL actually takes off, or something I may just write myself), which would be _intended_ to enforce a particular interpretation and structure for these values. As far as KDL itself is concerned, `<1>` and `1` are actually different types.

For your second example, that's a simpler answer: `my-node` is an empty node with no values, and `my-node <>` is a node with a single value, "none".

A JavaScript deserialization of the first might look like: {name: "my-node", values: [], props: {}, children: []}`, whereas the second would look like `{name: "my-node", values: [null], props: {}, children: []}`. It may end up _behaving_ the same in practice, but that depends on whether the program that's consuming this data treats `null` as `undefined`, or cares about the `values` length.

**Reply by lulu-berlin** [reply]
So you leave the decision to the interpretation done when deserializing. In this sense, maybe
```
my-node <123>
```
is too explicit, since it can be shortened to
```
my-node 123
```
and let the magic happen in deserialization.

In this case, what's left of the proposal is that `<>` could still be a nice short way to write `null`.

**Reply by zkat** [reply]
Yup, you got it :)

**Reply by lulu-berlin** [reply]
Actually my previous response didn't do justice to your comment. You actually said that you'd like to have both ways: explicit typed values, or concise values that let the deserialization worry about the types.

So the question now is: what's the advantage of having both ways?

**Reply by zkat** [reply]
KDL itself is based around concise values and lets the deserialization figure out its actual concrete types. It puts very few requirements on how its types should be represented.

HOWEVER, I want KDL to be designed to _allow_ explicitly-statically-typed values, and I think having special syntax for None is going to help that.

**Reply by lulu-berlin** [reply]
And to try to answer myself: maybe it's about easier round-trips. If KDL can also be serialized from actual data programatically, then now the optionality of a value is explicit and human-readable (`MyStruct { value: Some(123) }` => `MyStruct "value"=<123>`). If some KDL documents are created by serialization, then there would be a use case where `my-node <123>` is preferred to `my-node 123`, to reduce the complexity of deserialization.

**Reply by lilyball** [reply]
I’m not sure how this reduces complexity. If anything, it increases complexity, because now instead of just having to handle “is it null, a number, or other?” the `MyStruct` decoder has to handle “is it a null, a number, a &lt;number&gt;, or other?”.

**Reply by lulu-berlin** [reply]
In the example I gave, the deserializer can make a strong assumption that the property `value` of `MyStruct` is always an optional number (`<number>` in your notation). In the rust use case, this can be automatically derived from the type with a macro.

[This entails that any other type appearing in the document will yield a parse error.]

**Reply by lilyball** [reply]
Right, which means the deserializer would have to treat `number` and `<number>` identically.

**Reply by lulu-berlin** [reply]
No. It means `number` is an invalid value in this use-case. 

`MyStruct "value"=<>` => `Ok(myStruct { value: None })`
`MyStruct "value"=<123>` => `Ok(myStruct { value: Some(123) })`
`MyStruct "value"=123` => `Err(ParseError)`

**Reply by lulu-berlin** [reply]
It means you can have a strict deserializer that rejects non-conforming values, or you can go duck-typed and KDL supports both use-cases, depending on what makes more sense for the problem that's being solved with it.

**Reply by lilyball** [reply]
I would consider that to be a buggy implementation. There’s no benefit to be gained from turning something that has a very clear meaning into an error. `Option<i64>` is a union type equivalent to `number | null`, which is to say it’s inhabited either by a number or by null.

**Reply by lilyball** [reply]
~~“Oh sorry, you gave me a number, but I was expecting a construct where you told me explicitly that you had a number that wasn’t null”.~~ I’ve never before heard of a textual serialization format that worked this way. And even strictly-typed programming languages don’t necessarily work this way. For example, Swift will “hoist” any value into an optional (or an optional optional, or optional optional optional, etc) as needed to typecheck. so I could say `MyStruct(value: 123)` and this works just as well as `MyStruct(value: .some(123))`.

**Reply by lulu-berlin** [reply]
@lilyball I feel ridiculed by your comments. It definitely doesn't encourage me to participate in this discussion and try to share my perspective. I'm going to check out from this conversation.

**Reply by lilyball** [reply]
I apologize, I’m not trying to ridicule you. I was attempting to illustrate why `<123>` and `123` must be treated the same by any code that deserializes into typed structs.

Edit: To be more clear, this is entirely my fault. I understand why you interpreted my comment that way, I phrased it poorly. I was trying to anthropomorphize a parser but I did so without considering how this may come across to the reader.

**Reply by lilyball** [reply]
I value your input and the last thing I want to do is drive you (or anyone else) away.

**Comment by lilyball** [comment]
KDL doesn’t have static typing. Given that, I don’t think there’s any value to be gained in trying to mark a value as `Some(1)`. The fact that it’s an integer instead of `null` already provides this information. The only practical difference here is in the ability to represent something like `Some(None)`, and I’m not convinced that’s worth adding this.

If you think about it, an optional type like `Option<i64>` is really just a union type like `number | null`. Dynamic typing already works with union types, so any hypothetical KDL schema language would likely just use union types to describe the options for a given value/property.

**Reply by zkat** [reply]
My thinking here is more about languages that want to be more explicit about it (like loose deserializations in Rust), but also when there comes a time when schema tools for KDL exist, and those would have more powerful tools to play with.

In the meantime, Just (hah) using `<>` for `null` and unwrapped values for not-null values will be good enough for most people. I just want this handy.

**Reply by lilyball** [reply]
I’m not convinced this is particularly useful for schemas. `<1>` and `<<1>>` and `<<<1>>>` are semantically equivalent to `1`. The only distinction is `<<>>`, `<<<>>>`, etc and that seems like such an edge case. And this means “using up” angle brackets, preventing their use for some other feature.

**Reply by jlkiri** [reply]
I would agree with @lilyball and feel that this

>The fact that it’s an integer instead of null already provides this information

is the biggest point. Although it's a different story if KDL itself will be used as a schema language. In JSON Schema (which I'm not very familiar with) it seems you can express a nullable string like this:

```
{ "type": ["string", "null"] }
```

which is much more verbose than `type <"string">`.

**Reply by lilyball** [reply]
Presumably that same syntax lets you write `{"type": ["string", "number"]}` too, which is to say it’s just defining a union type.
